weld/subsystem/src/main/java/org/jboss/as/weld/services/bootstrap/WeldExecutorServices.java

```diff
@@ -115,4 +118,52 @@ public synchronized ExecutorService getTaskExecutor() {
     public void cleanup() {
         // noop on undeploy - the executor is a service shared across multiple deployments
     }
+
+    class WeldExecutor extends ThreadPoolExecutor {
+
+        WeldExecutor(int nThreads, ThreadFactory threadFactory) {
+            super(nThreads, nThreads,
+                               0L, TimeUnit.MILLISECONDS,
+                               new LinkedBlockingQueue<Runnable>(),
+                               threadFactory);
+        }
+
+        @Override
+        protected void beforeExecute(Thread t, Runnable r) {
+            super.beforeExecute(t, r);
+            if (r instanceof WeldTaskWrapper) {
+                NamespaceContextSelector.pushCurrentSelector(((WeldTaskWrapper) r).currentSelector);
+            }
+        }
+
+        @Override
+        protected void afterExecute(Runnable r, Throwable t) {
+            super.afterExecute(r, t);
+            if (r instanceof WeldTaskWrapper) {
+                NamespaceContextSelector.popCurrentSelector();
+            }
+        }
+
+        @Override
+        public void execute(Runnable command) {
+            WeldTaskWrapper task = new WeldTaskWrapper(command, NamespaceContextSelector.getCurrentSelector());
```

Comment:
Avoid wrapping if Weld isn't yet started, `org.jboss.weld.Container#instance()` can be used to check state.
You can see how we do similar check here - https://github.com/wildfly/wildfly/blob/master/weld/subsystem/src/main/java/org/jboss/as/weld/WeldProvider.java#L69-L70

Reply:
Thanks @manovotn, I added the check